packages/location-management/src/components/LocationTree/index.tsx

```diff
@@ -1,110 +1,92 @@
-  reducer,
-} from '../../ducks/location-hierarchy';
 import { AntTreeProps } from '../LocationUnitList';
 import './tree.css';
 import { ParsedHierarchyNode } from '../../ducks/locationHierarchy/types';
-import { getHierarchyNodeFromArray } from '../../ducks/locationHierarchy/utils';
 import { useTranslation } from '../../mls';
+import TreeModel from 'tree-model';
 import { Key } from 'rc-tree/lib/interface';
-reducerRegistry.register(reducerName, reducer);
+
+/** helper type, shortened form */
+export type TreeNode = TreeModel.Node<ParsedHierarchyNode>;
+
+interface SelectCallbackInfo {
+  event: 'select';
+  selected: boolean;
+  node: TreeNode;
+  selectedNodes: TreeNode;
+  nativeEvent: MouseEvent;
+}
 
 interface TreeProp {
-  data: ParsedHierarchyNode[];
-  OnItemClick: (item: ParsedHierarchyNode) => void;
+  data: TreeNode[];
+  selectedNode: TreeNode | undefined;
+  onSelect: (item: TreeNode | undefined) => void;
+}
+
+interface ExpandCallbackInfo {
+  node: TreeNode;
+  expanded: boolean;
+  nativeEvent: MouseEvent;
 }
 
 export const Tree: React.FC<TreeProp> = (props: TreeProp) => {
-  const { data, OnItemClick } = props;
-  const dispatch = useDispatch();
+  const { data, onSelect, selectedNode } = props;
   const { t } = useTranslation();
 
   const [expandedKeys, setExpandedKeys] = useState<Key[]>([]);
   const [searchValue, setSearchValue] = useState<string>('');
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  const filterData: ParsedHierarchyNode[] = [];
-  const locationTreeState = useSelector((state) => getLocationTreeState(state));
-
-  /**
-   * Function to to expand tree
-   *
-   * @param {string} value - value to be searched and expanded
-   */
-  function expandTree(value: string) {
-    const expandedKeys = filterData
-      .map((item) =>
-        value.length && item.label.toLocaleLowerCase().indexOf(value.toLowerCase()) > -1
-          ? getParentKey(item.id, item.parent as string, filterData)
-          : null
-      )
-      .filter((item, i, self) => item && self.indexOf(item) === i);
-    setExpandedKeys(expandedKeys as string[]);
-    dispatch(
-      setLocationTreeState({
-        keys: expandedKeys as React.Key[],
-        node: locationTreeState?.node as ParsedHierarchyNode,
-      })
-    );
-  }
+  const [searchExpandedKeys, setSearchExpandedKeys] = useState<Key[]>([]);
 
   useEffect(() => {
-    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-    if (locationTreeState?.node) {
-      const node = getHierarchyNodeFromArray(filterData, locationTreeState.node.id);
-      if (node) {
-        onExpand(locationTreeState.keys);
-        OnItemClick(node);
-      }
+    const updateExpandedKeys = (keys: Key[]) => setExpandedKeys([...expandedKeys, ...keys]);
+    // set expanded and selected keys depending on location tree state and searchValue
+    if (selectedNode) {
+      let parentPath = selectedNode.getPath();
+      parentPath = parentPath.slice(1, parentPath.length);
```

Comment:
could we avoid in place mutation here?

Reply:
also are we okay with shallow copies here?

Reply:
actually what are we trying to achieve here?

Reply:
> actually what are we trying to achieve here?

get the keys for the parent nodes to the selected node.

> also are we okay with shallow copies here?

yeah its okay for this case. The tree data is not edited, wrapping it in an immutable wrapper like Immutable js would still yield the same results.

> could we avoid in place mutation here?

`getPath` will include the node itself as part of the parentPath, but we are only interested in the parent nodes only. I might be missing something, but I think this is perfectly valid code.

Reply:
I'm asking for us to have separate variables, not necessarily an immutable wrapper. It's an anti pattern to mutate values in place

```
     const somePath = selectedNode.getPath();
     const parentPath = somePath.slice(1, somePath.length);
```

Reply:
> It's an anti pattern to mutate values in place

Do you have sources on this, would that mean there is no proper case for having/using `let` since the whole goal of such a language construct is to allow programmers to perform such mutations.